hathor/p2p/sync_v2/mempool.py

```diff
@@ -0,0 +1,116 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import deque
+from typing import TYPE_CHECKING, Any, Generator, Optional
+
+from structlog import get_logger
+from twisted.internet.defer import Deferred, inlineCallbacks
+
+from hathor.transaction import BaseTransaction
+
+if TYPE_CHECKING:
+    from hathor.p2p.sync_v2.manager import NodeBlockSync
+
+logger = get_logger()
+
+
+class SyncMempoolManager:
+    """Manage the sync-v2 mempool with one peer.
+    """
+    def __init__(self, sync_manager: 'NodeBlockSync'):
+        """Initialize the sync-v2 mempool manager."""
+        self.log = logger.new(peer=sync_manager.protocol.get_short_peer_id())
+
+        # Shortcuts.
+        self.sync_manager = sync_manager
+        self.manager = self.sync_manager.manager
+        self.tx_storage = self.manager.tx_storage
+        self.reactor = self.sync_manager.reactor
+
+        # Set of tips we know but couldn't add to the DAG yet.
+        self.missing_tips: set[bytes] = set()
+
+        # Maximum number of items in the DFS.
+        self.MAX_STACK_LENGTH: int = 1000
+
+        # Whether the mempool algorithm is running
+        self._is_running = False
+
+    def is_running(self) -> bool:
+        """Whether the sync-mempool is currently running."""
+        return self._is_running
+
+    def run(self) -> None:
+        """Starts _run in, won't start again if already running."""
+        if self.is_running():
+            self.log.warn('already started')
+        else:
+            self._is_running = True
+            self.reactor.callLater(0, self._run)
+
+    @inlineCallbacks
+    def _run(self) -> Generator[Deferred, Any, None]:
+        """Run a single loop of the sync-v2 mempool."""
+        if not self.missing_tips:
+            # No missing tips? Let's get them!
+            tx_hashes: list[bytes] = yield self.sync_manager.get_tips()
+            self.missing_tips.update(h for h in tx_hashes if not self.tx_storage.transaction_exists(h))
+
+        while self.missing_tips:
+            self.log.debug('We have missing tips! Let\'s start!', missing_tips=[x.hex() for x in self.missing_tips])
+            tx_id = next(iter(self.missing_tips))
+            tx: BaseTransaction = yield self.sync_manager.get_tx(tx_id)
+            # Stack used by the DFS in the dependencies.
+            # We use a deque for performance reasons.
+            self.log.debug('start mempool DSF', tx=tx.hash_hex)
+            yield self._dfs(deque([tx]))
+
+        # sync_manager.run_sync will start it again when needed
+        self._is_running = False
```

Comment:
We need to add protection here. If an exception is raised in this method, sync will stop running because this flag will always be true.

Reply:
Done!